facts: fix double-counting of CPUs on POWER systems

yselkowitz opened the pull request:
On POWER systems, /proc/cpuinfo provides a 'processor' entry as a
counter, and a 'cpu' entry with a description (similar to 'model name'
on x86). Support for POWER in get_cpu_facts was added via the 'cpu'
entry in commit 8746e692c121d7081e5c545de8e98908a95b510b.  Subsequent
support for ARM64 in commit ce4ada93f9daae4e1806c612b5c2f96727dc3de5
used the 'processor' entry, resulting in double-counting of cores on
POWER systems.

Signed-off-by: Yaakov Selkowitz <yselkowi@redhat.com>

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
facts

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
On a ppc64le node, `ansible_processor_vcpus` currently shows ansible_processor_vcpus twice the number listed in `/proc/cpuinfo`.  With this patch, the count matches.

<!--- Paste verbatim command output below, e.g. before and after your change -->
Before:
```
$ ansible all -v -m setup -a 'filter=ansible_processor_vcpus'
Using /etc/ansible/ansible.cfg as config file
localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_processor_vcpus": 352, 
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "changed": false
}

$ grep '^processor[[:space:]]*:' /proc/cpuinfo
processor	: 0
[snip]
processor	: 175
```

After:
```
$ ansible all -v -m setup -a 'filter=ansible_processor_vcpus'
Using /etc/ansible/ansible.cfg as config file
localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_processor_vcpus": 176, 
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "changed": false
}
```

yselkowitz commented:
We would like to see this merged to supported stable branches as well.

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/129700/9/tests

yselkowitz commented:
The problem was in the test, just pushed a new patch to fix that as well.

yselkowitz commented:
0a922a1 fixes the unit tests, I'm not sure what the failure in T=linux/fedora29/4 was.

yselkowitz commented:
All checks have passed now; the issue with linux/fedora29/4 must have been unrelated.  Anything else needed to get this merged?

yselkowitz commented:
/ready_for_review

yselkowitz commented:
/bot_status

yselkowitz commented:
ready_for_review

samdoran commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -204,7 +204,7 @@ def get_cpu_facts(self, collected_facts=None):
             # works for some ARM devices, like the Sheevaplug.
             # 'ncpus active' is SPARC attribute
-            if key in ['model name', 'Processor', 'vendor_id', 'cpu', 'Vendor', 'processor']:
+            if key in ['model name', 'Processor', 'vendor_id', 'Vendor', 'processor']:
```
I'm hesitant to remove the `cpu` flag. It does pass our unit tests, but there may be architectures out there relying on `cpu` as a counter that we don't have fixtures for yet.

yselkowitz replied:
The history in the commit message shows that `cpu` was added explicitly for Power.  If something else is relying thereon (and it's certainly not x86_64, s390x, aarch64, and apparently not sparc64), then it's probably a fluke that it works at all.

samdoran commented on test/units/module_utils/facts/hardware/linux_data.py:
```diff
@@ -484,65 +484,33 @@
-                '6', 'POWER7 (architected), altivec supported',
-                '7', 'POWER7 (architected), altivec supported'
-            ],
+            'processor': ['0', '1', '2', '3', '4', '5', '6', '7'],
```
Changing these fixtures is incorrect. I get test failures just having a list of integers since the output is:

```
cpu_facts['processor']=[
    '0', 'POWER7 (architected), altivec supported', 
    '1', 'POWER7 (architected), altivec supported', 
    '2', 'POWER7 (architected), altivec supported', 
    '3', 'POWER7 (architected), altivec supported', 
    '4', 'POWER7 (architected), altivec supported', 
    '5', 'POWER7 (architected), altivec supported', 
    '6', 'POWER7 (architected), altivec supported', 
    '7', 'POWER7 (architected), altivec supported'
]
```

yselkowitz replied:
Did you get those outputs with my change or with your suggestion?

samdoran replied:
This was with my suggestion. That would explain the difference.

samdoran commented on test/units/module_utils/facts/hardware/linux_data.py:
```diff
@@ -484,65 +484,33 @@
-                '23', 'POWER8 (architected), altivec supported',
+                '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
+                '11', '12', '13', '14', '15', '16', '17', '18', '19', '20',
+                '21', '22', '23',
```
Same as above. I don't believe these fixtures should change.

samdoran reviewed (changes requested):
I think a better solution is to use the same code that fixes the double processoring counting for ARM CPUs:

```patch
diff --git a/test/units/module_utils/facts/hardware/test_linux_get_cpu_info.py b/test/units/module_utils/facts/hardware/test_linux_get_cpu_info.py
index 542ed4bce2..993f5f1219 100644
--- a/test/units/module_utils/facts/hardware/test_linux_get_cpu_info.py
+++ b/test/units/module_utils/facts/hardware/test_linux_get_cpu_info.py
@@ -32,7 +32,7 @@ def test_get_cpu_info_missing_arch(mocker):
     for test in CPU_INFO_TEST_SCENARIOS:
         mocker.patch('ansible.module_utils.facts.hardware.linux.get_file_lines', side_effect=[[], test['cpuinfo']])
         test_result = inst.get_cpu_facts()
-        if test['architecture'].startswith(('armv', 'aarch')):
+        if test['architecture'].startswith(('armv', 'aarch', 'ppc')):
             assert test['expected_result'] != test_result
         else:
             assert test['expected_result'] == test_result
```

yselkowitz committed:
facts: fix double-counting of CPUs on POWER systems

On POWER systems, /proc/cpuinfo provides a 'processor' entry as a
counter, and a 'cpu' entry with a description (similar to 'model name'
on x86). Support for POWER in get_cpu_facts was added via the 'cpu'
entry in commit 8746e692c121d7081e5c545de8e98908a95b510b.  Subsequent
support for ARM64 in commit ce4ada93f9daae4e1806c612b5c2f96727dc3de5
used the 'processor' entry, resulting in double-counting of cores on
POWER systems.

When unit tests were later written for this code in
commit 55306906cf7577cc5a666bf6f3d0486dcf937c03, the erroneous values
were just accepted in the test instead of being diagnosed.

Signed-off-by: Yaakov Selkowitz <yselkowi@redhat.com>

yselkowitz commented:
@samdoran a new patch has been posted.  Looks like CI needs a rekick for T=rhel/8.0/2 only.

samdoran commented:
rebuild_merge

yselkowitz commented:
How do I relaunch CI to deal with what appears to be a fluke w/o losing the approvals?

samdoran commented:
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs) for this to be included in previous versions.
